src/parser/core/SharedCode.tsx

```diff
@@ -0,0 +1,231 @@
+import EventHistory from 'parser/shared/modules/EventHistory';
+import SpellUsable from 'parser/shared/modules/SpellUsable';
+
+class SharedCode extends Analyzer {
+  static dependencies = {
+    abilityTracker: AbilityTracker,
+    spellUsable: SpellUsable,
+    enemies: Enemies,
+    eventHistory: EventHistory,
+  };
+  protected abilityTracker!: AbilityTracker;
+  protected spellUsable!: SpellUsable;
+  protected enemies!: Enemies;
+  protected eventHistory!: EventHistory;
+
+  /**
+   * @param cast the cast event that should be evaluated
+   * @returns a true/false of whether the precast was a hardcast or not
+   */
+  isHardcast(cast: CastEvent) {
+    const beginCast = this.getEvents(
+      true,
+      EventType.BeginCast,
+      SPELLS[cast.ability.guid],
+      1,
+      cast.timestamp,
+    )[0];
+    return cast.timestamp - beginCast.timestamp > 50;
+  }
+
+  /**
+   * @param event the event you want to get the pre cast for
+   * @param preCastSpell The spell object of the specific spell (or an array of spells) you want to check for
+   * @returns a true/false of whether the precast was found or not.
+   */
+  checkPreCast(event: AnyEvent, preCastSpell: SpellInfo | SpellInfo[]) {
+    const preCast = this.getPreCast(event, preCastSpell);
+    return preCast ? true : false;
+  }
+
+  /**
+   * @param event the event you want to get the pre cast for
+   * @param preCastSpell an optional spell that you want to look for. Leave undefined to get the precast spell regardless of what spell it is.
+   * @returns the event for the precast spell, or undefined if none was found.
+   */
+  getPreCast(event: AnyEvent, preCastSpell?: SpellInfo | SpellInfo[]) {
+    return this.getEvents(true, EventType.Cast, preCastSpell, 1, event.timestamp, 250)[0];
+  }
+
+  /**
+   * @param eventType The type of event you want to count. i.e. EventType.Cast, EventType.BeginCast, etc. Use EventType.Event for all events.
+   * @param spell the spell you want to count casts for.
+   * @returns the number of events found
+   */
+  countEvents(eventType: EventType, spell: SpellInfo) {
+    const events = this.getEvents(true, eventType, spell);
+    return events.length;
+  }
+
+  /**
+   * @param buffActive filters based on whether the buff is active or inactive (true for active, false for inactive)
+   * @param buff the spell object for the buff
+   * @param eventType the type of event that you want to search for. i.e. EventType.Cast, EventType.BeginCast, etc. Use EventType.Event for all events.
+   * @param cast an optional cast spell object to count. Omit or leave undefined to count all casts
+   * @returns the number of events found
+   */
+  countEventsByBuff(buffActive: boolean, buff: SpellInfo, eventType: EventType, cast?: SpellInfo) {
+    const events = buffActive
+      ? this.getEventsByBuff(true, buff, eventType, cast)
+      : this.getEventsByBuff(false, buff, eventType, cast);
+    return events.length;
+  }
```

Comment:
generally, I don't think we need additional methods for `countX` when the body is (almost) just `X.length`.

in the case of `countEventsByBuff``, you will only ever call it with `buffActive` being true (if your code cares about events while the buff was up) or false (for the opposite)---it will never vary from call to call, so the ternary won't come up in caller code.